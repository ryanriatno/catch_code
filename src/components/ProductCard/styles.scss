.product {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: space-around;
  text-align: center;
  width: 100%;
  padding-bottom: 3em;
  &__price {
    position: absolute;
    bottom: 1em;
    left: 0;
    width: 100%;
    font-weight: bold;
  }
  &__sale-price {
    text-decoration: line-through;
    display: block;
    margin: 1em 0 3em 0;
  }
  &__lists {
    background: #fff;
    position: relative;
    flex: 0 0 10%;
    border: 1px solid #ddd;
    padding: 10px 5px;
    margin: 5px;
    min-width: 230px;
    &-price {
      margin: 1em 0;
      display: block;
    }
    &-cover {
      position: relative;
    }
    &-sold-out {
      background: red;
      color: white;
      font-size: 14px;
      position: absolute;
      top: 0;
      left: 0;
      padding: .5em 1em;
      font-weight: bold;
    }
    &-name {
      height: 48px;
      overflow: hidden;
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;  
    }
    &-detail {
      padding: 1em;
      text-align: center;
    }
  }
}