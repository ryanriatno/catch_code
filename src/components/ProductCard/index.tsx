import React, { useState, useEffect } from 'react'
import './styles.scss'

interface Props {
  imageUrl: string
  salePrice: number
  retailPrice: number
  id: string|number
  quantityAvailable: number
  name: string
}
const ProductCard: React.FC<Props> = (props) => {
  const [onSales, setOnSales] = useState(false)
  const convertCurrency = (value) => {
    let price = value / 100
    return price.toLocaleString('en-US', { style:'currency', currency:'USD' })
  }

  useEffect(() => {
    if (props.retailPrice > 0) {
      setOnSales(true)
    }else{
      setOnSales(false)
    }
  }, [props.retailPrice])

  return (
    <div className="product__lists" key={props.id}>
      <div className="product__lists-cover">
        <img src={props.imageUrl} alt="" />
        {(props.quantityAvailable < 1) ? <span data-testid="soldout" className="product__lists-sold-out">SOLD OUT</span> : null}
      </div>
      <div className="product__lists-detail">
        <h4 className="product__lists-name">{props.name}</h4>
        {
          (onSales) ?
            <span data-testid="retail-price" className="product__sale-price">{convertCurrency(props.retailPrice)}</span>
            : null
        }
        {
          (onSales) ?
          <span data-testid="price" className="product__price">{convertCurrency(props.salePrice)}</span>
          :
          <span data-testid="price" className="product__price">{convertCurrency(props.retailPrice)}</span>
        }
      </div>
    </div>
  )
}

export default ProductCard