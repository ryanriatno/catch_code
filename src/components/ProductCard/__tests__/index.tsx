import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import ProductCard from '../index'
import {
  getByTestId,
} from '@testing-library/dom'

describe("<ProductCard /> test case", () => {
  test("Render Header without error", () => {
    const data = {
      quantityAvailable: 1,
      key:"123",
      name:"Shoe",
      imageUrl:"https://s.catch.com.au/images/product/0001/1909/5d47c0d64988e613254534_w200.jpg",
      salePrice:1234,
      retailPrice:4321,
      id:123123123
    }
    const { container } = render(<ProductCard {...data} />);
    expect(container).toMatchSnapshot();
  });
  test("Display SOLD OUT if quantity is 0", () => {
    const data = {
      quantityAvailable: 0,
      key:"123",
      name:"Shoe",
      imageUrl:"https://s.catch.com.au/images/product/0001/1909/5d47c0d64988e613254534_w200.jpg",
      salePrice:1234,
      retailPrice:4321,
      id:123123123
    }
    const { container } = render(<ProductCard {...data} />);
    expect(getByTestId(container, 'soldout')).toHaveTextContent(
      'SOLD OUT'
    )
  })
  test("Show sale price correctly", () => {
    const data = {
      quantityAvailable: 0,
      key:"123",
      name:"Shoe",
      imageUrl:"https://s.catch.com.au/images/product/0001/1909/5d47c0d64988e613254534_w200.jpg",
      salePrice:1234,
      retailPrice:4321,
      id:123123123
    }
    const { container } = render(<ProductCard {...data} />);
    expect(getByTestId(container, 'price')).toHaveTextContent(
      '$12.34'
    )
  })
});