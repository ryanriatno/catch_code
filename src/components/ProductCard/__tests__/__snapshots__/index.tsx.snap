// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> test case Render Header without error 1`] = `
<div>
  <div
    class="product__lists"
  >
    <div
      class="product__lists-cover"
    >
      <img
        alt=""
        src="https://s.catch.com.au/images/product/0001/1909/5d47c0d64988e613254534_w200.jpg"
      />
    </div>
    <div
      class="product__lists-detail"
    >
      <h4
        class="product__lists-name"
      >
        Shoe
      </h4>
      <span
        class="product__sale-price"
        data-testid="retail-price"
      >
        $43.21
      </span>
      <span
        class="product__price"
        data-testid="price"
      >
        $12.34
      </span>
    </div>
  </div>
</div>
`;
